src/hooks/useJournalData.ts

```diff
@@ -0,0 +1,252 @@
+    client?: JournalClient;
+    documentCount: number;
+}) {
+    if (!client || !journalName) {
+        console.warn('Cannot load documents without client and journal');
+        return [];
+    }
+    const metaInfo = (
+        await client.read({
+            metadataOnly: true,
+            journal: journalName,
+        })
+    ).unwrap();
+
+    const generator = streamAsyncIterator(metaInfo);
+
+    const metadataResponse = (await generator.next()).value;
+
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    if (metadataResponse?.writeHead) {
+        const head = parseInt(metadataResponse.writeHead, 10);
+        let start = head;
+
+        let documents: JournalRecord[] = [];
+
+        while (documents.length < documentCount && start > 0) {
+            start = Math.max(0, start - INCREMENT);
+            const stream = (
+                await client.read({
+                    journal: journalName,
+                    offset: `${start}`,
+                    endOffset: `${head}`,
+                })
+            ).unwrap();
+            const journalDocumentStream = parseJournalDocuments(stream);
+            const allDocs = await readAllDocuments(journalDocumentStream);
+
+            // TODO: Instead of inefficiently re-reading until we get the desired row count,
+            // we should accumulate documents and shift `head` backwards using `ProtocolReadResponse.offset`
+            documents = allDocs
+                .filter(isJournalRecord)
+                .filter(
+                    (record) =>
+                        !(record._meta as unknown as { ack: boolean }).ack
+                )
+                .slice(documentCount * -1);
+        }
+        return documents;
+    } else {
+        throw new Error('Unable to load metadata');
+    }
+}
+
+export type JournalRecord<B extends {} = Record<string, any>> = B & {
+    _meta: {
+        uuid: string;
+    };
+};
+
+function isJournalRecord(val: any): val is JournalRecord {
+    return val?._meta?.uuid;
+}
+
+const useJournalData = (journalName?: string, desiredCount: number = 50) => {
+    const [gatewayConfig] = useLocalStorage(
+        LocalStorageKeys.GATEWAY,
+        getStoredGatewayAuthConfig()
+    );
+
+    const journalClient = useMemo(() => {
+        if (journalName && gatewayConfig?.gateway_url && gatewayConfig.token) {
+            const authToken = gatewayConfig.token;
+            const baseUrl = new URL(gatewayConfig.gateway_url);
+
+            return new JournalClient(baseUrl, authToken);
+        } else {
+            return undefined;
+        }
+    }, [gatewayConfig, journalName]);
+
+    const [refreshCount, setRefresh] = useState(0);
+
+    const [data, setData] = useState<JournalRecord[]>([]);
+    const [error, setError] = useState<any>(null);
+    const [loading, setLoading] = useState(false);
+
+    useEffect(() => {
+        void (async () => {
+            if (journalName && journalClient) {
```

Comment:
It seems a little weird to me that we handle the case where `journalName` is null all the way down here. Is there a reason to? My normal intuition (which is often wrong WRT front-end) is that `journalName` should be a required parameter, and the caller should handle the case where there are no journals for a given collection.

Reply:
So this is a weird react hooks thing (which I'm curious about your opinion on @travjenkins). Because you can't conditionally call a hook, the question is what to do when (in this case) the journal info hasn't loaded yet. My solution was to allow passing null/undefined for `journalName`, and if it's not provided, then don't attempt to fetch the data, and just return an empty array.

Reply:
Yup... this is just kinda how it is done. Also, we tend to put all our effects near the bottom (as in right before the return) since a lot of times they need the variables and stuff declared above.